proposals/testing/NNNN-image-attachments-in-swift-testing-windows.md

```diff
@@ -0,0 +1,391 @@
+  /// see the documentation for the [`IWICBitmapEncoder`](https://learn.microsoft.com/en-us/windows/win32/api/wincodec/nn-wincodec-iwicbitmapencoder)
+  /// class.
+  public init(encoderCLSID: CLSID, encodingQuality: Float = 1.0)
+}
+```
+
+For convenience, an initializer is provided that takes a path extension and
+tries to map it to the appropriate codec's `CLSID` value:
+
+```swift
+extension AttachableImageFormat {
+  /// Construct an instance of this type with the given path extension and
+  /// encoding quality.
+  ///
+  /// - Parameters:
+  ///   - pathExtension: A path extension corresponding to the image format to
+  ///     use when encoding images.
+  ///   - encodingQuality: The encoding quality to use when encoding images. For
+  ///     the lowest supported quality, pass `0.0`. For the highest supported
+  ///     quality, pass `1.0`.
+  ///
+  /// If the target image format does not support variable-quality encoding,
+  /// the value of the `encodingQuality` argument is ignored.
+  ///
+  /// If `pathExtension` does not correspond to a recognized image format, this
+  /// initializer returns `nil`:
+  ///
+  /// - On Apple platforms, the content type corresponding to `pathExtension`
+  ///   must conform to [`UTType.image`](https://developer.apple.com/documentation/uniformtypeidentifiers/uttype-swift.struct/image).
+  /// - On Windows, there must be a corresponding subclass of [`IWICBitmapEncoder`](https://learn.microsoft.com/en-us/windows/win32/api/wincodec/nn-wincodec-iwicbitmapencoder)
+  ///   registered with Windows Imaging Component.
+  public init?(pathExtension: String, encodingQuality: Float = 1.0)
+}
+```
+
+For consistency, `init(pathExtension:encodingQuality:)` is provided on Apple
+platforms too. (This is the only part of this proposal that affects platforms
+other than Windows.)
+
+### Example usage
+
+A developer may then easily attach an image to a test by calling
+`Attachment.record()` and passing the image of interest. For example, to attach
+an icon to a test as a PNG file:
+
+```swift
+import Testing
+import WinSDK
+
+@MainActor @Test func `attaching an icon`() throws {
+  let hIcon: HICON = ...
+  defer {
+    DestroyIcon(hIcon)
+  }
+  Attachment.record(hIcon, named: "my icon", as: .png)
+  // OR: Attachment.record(hIcon, named: "my icon.png")
+}
+```
+
+## Source compatibility
+
+This change is additive only.
+
+## Integration with supporting tools
+
+None needed.
+
+## Future directions
+
+- Adding support for projecting COM classes as foreign-reference-counted Swift
+  classes. The C++ interop team is interested in implementing this feature, but
+  it is beyond the scope of this proposal. **If this feature is implemented in
+  the future**, it will cause types like `IWICBitmapSource` to be projected
+  directly into Swift instead of as `UnsafeMutablePointer` specializations. This
+  would be a source-breaking change for Swift Testing, but it would make COM
+  classes much easier to use in Swift.
```

Comment:
Do we have a clear idea of how we would evolve or adjust these proposed APIs to that change if/when it lands in the future?

Reply:
We don't have that information, no. Presumably a type like `IWICBitmap` would be imported directly into Swift instead of needing to be referenced via an unsafe pointer. But I don't know what specific changes we might need to make, no.

Reply:
Fair. Maybe a better question is: Do we have any reason to believe it may be problematic if/when that lands, i.e. break us in a way we cannot adapt to? As long as you don't have any specific reason to be worried about unfixable breakages, I'm comfortable

Reply:
I'm not aware of any specific issues at this time.